pkg/hintrunner/zero/zerohint_keccak.go

```diff
@@ -102,3 +105,93 @@ func createKeccakWriteArgsHinter(resolver hintReferenceResolver) (hinter.Hinter,
 
 	return newKeccakWriteArgsHint(inputs, low, high), nil
 }
+
+func newBlockPermutationHint(keccakStateSizeFelts, keccakPtr hinter.ResOperander) hinter.Hinter {
+	name := "BlockPermutation"
+	return &GenericZeroHinter{
+		Name: name,
+		Op: func(vm *VM.VirtualMachine, _ *hinter.HintRunnerContext) error {
+			// from starkware.cairo.common.keccak_utils.keccak_utils import keccak_func
+			// _keccak_state_size_felts = int(ids.KECCAK_STATE_SIZE_FELTS)
+			// assert 0 <= _keccak_state_size_felts < 100
+
+			// output_values = keccak_func(memory.get_range(
+			// 	ids.keccak_ptr - _keccak_state_size_felts, _keccak_state_size_felts))
+			// segments.write_arg(ids.keccak_ptr, output_values)
+
+			keccakStateSize, err := hinter.ResolveAsUint64(vm, keccakStateSizeFelts)
+			if err != nil {
+				return err
+			}
+
+			keccakWritePtr, err := hinter.ResolveAsAddress(vm, keccakPtr)
+			if err != nil {
+				return err
+			}
+
+			// why does the need contains this check? keccakStateSize is a constant: 25
+			if keccakStateSize >= 100 {
+				return fmt.Errorf("keccakStateSize %v is out range 0 <= keccakStateSize < 100", &keccakStateSize)
+			}
+
+			var readAddr = *keccakWritePtr
+			var offset int16 = int16(keccakStateSize)
+			var negOffset int16 = -offset
+
+			readAddr, err = readAddr.AddOffset(negOffset)
+			if err != nil {
+				return err
+			}
+
+			inputValuesInRange, err := hinter.GetConsecutiveValues(vm, readAddr, int16(keccakStateSize))
+			if err != nil {
+				return err
+			}
+
+			var keccakInput []uint64
+
+			for _, valueMemoryValue := range inputValuesInRange {
+				valueFelt, err := valueMemoryValue.FieldElement()
+				if err != nil {
+					return err
+				}
+				valueUint256 := uint256.Int(valueFelt.Bits())
+				valueUint64 := valueUint256.Uint64()
+				keccakInput = append(keccakInput, valueUint64)
+			}
+
+			var input [25]uint64
+
+			for i := 0; i < 25; i++ {
+				input[i] = keccakInput[i]
+			}
+			builtins.KeccakF1600(&input)
```

Comment:
When looking at this something crossed my mind. I was using this function with an array, which contained _keccak_state_size_felts * [0]. So the _keccak_state_size_felts could be 1, could be 99, the result was always the same as for fixed size keccak_input[25]. Assuming, that __keccak_state_size_felts can be different than 25, did you check that the output is the same? For instance, if the __keccak_state_size_felts is 30, in the loop above you are making a cutoff at 25. I'm just curious. 

We have to investigate, if the __keccak_state_size_felts can be different than 25, because if so, implementing the keccak_func by hand might be neccessary.

Reply:
keccakF1600 takes `*[25]uint64` as input, i cannot be > 25

are you talking about CairoKeccakFinalize ? I just look at it your code is : 

```
var input [25]uint64
builtins.KeccakF1600(&input)
```

you pass an array of 25 uint64. I think it must always be 25 elements

Reply:
by the way, for CairoKeccakFinalize, the python code is `keccak_func(inp)` with `inp = [0] * _keccak_state_size_felts`

is it correct to apply KeccakF1600 to an array of 25 elements set to 0 ? what is `[0]` ?

Reply:
I think that because `KeccakF1600` necessarily takes a ` [25]uint64` , indeed you can have only 1 or 2 values set to a non zero value and you just pass an array of 25 elements with all other values set to 0. But the keccak state cannot be > 25 I suppose as you just couldnt pass it to `KeccakF1600`

Reply:
Regarding your previous comment, I have checked the keccak_func(inp) for different length of arrays filled with zero, and the output was always the same as for the array of length 25, so that is not a prolem.

Reply:
> by the way, for CairoKeccakFinalize, the python code is `keccak_func(inp)` with `inp = [0] * _keccak_state_size_felts`
> 
> is it correct to apply KeccakF1600 to an array of 25 elements set to 0 ? what is `[0]` ?

It is just array of __keccak_state_size_felts filled with zeros

Reply:
> I think that because `KeccakF1600` necessarily takes a ` [25]uint64` , indeed you can have only 1 or 2 values set to a non zero value and you just pass an array of 25 elements with all other values set to 0. But the keccak state cannot be > 25 I suppose as you just couldnt pass it to `KeccakF1600`

this seems legit, I dont see any overwriting of the 25 const, so you are right

Reply:
oh yes ok great i'm not used to go syntax sometimes i get confused
perfect!